src/Analyzer.JsonRuleEngine/Operators/InequalityOperator.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Templates.Analyzer.RuleEngines.JsonEngine.Operators
+{
+    /// <summary>
+    /// An operator that evaluates the "greater", "greaterOrEquals", "less", and "lessOrEquals" JSON expressions.
+    /// </summary>
+    internal class InequalityOperator : LeafExpressionOperator
+    {
+        /// <inheritdoc/>
+        public override string Name => GetName();
+
+        /// <summary>
+        /// Whether the operator also considers equality
+        /// </summary>
+        public Boolean OrEquals;
+
+        /// <summary>
+        /// Creates an InequalityOperator.
+        /// </summary>
+        /// <param name="specifiedValue">The value specified in the JSON rule.</param>
+        /// <param name="isNegative">Whether the operator compares by greater than or by less than.</param>
+        /// <param name="orEquals">Whether the operator also considers equality.</param>
+        public InequalityOperator(JToken specifiedValue, bool isNegative, bool orEquals)
+        {
+            if (specifiedValue == null)
+            {
+                throw new ArgumentNullException(nameof(specifiedValue));
+            }
+            else
+            {
+                ValidateComparisonTerm(specifiedValue);
+            }
+
+            this.SpecifiedValue = specifiedValue;
+            this.IsNegative = isNegative;
+            this.OrEquals = orEquals;
+        }
+
+        /// <summary>
+        /// Evaluates <paramref name="tokenToEvaluate"/> to determine if it is greater, greaterOrEquals, less, or lessOrEquals than the specified value.
+        /// </summary>
+        /// <param name="tokenToEvaluate">The JToken to evaluate.</param>
+        /// <returns>A value indicating whether or not the evaluation passed.</returns>
+        public override bool EvaluateExpression(JToken tokenToEvaluate)
+        {
+            if (tokenToEvaluate == null)
+            {
+                // If the specified property in the JSON is not defined then we would assume it could potentially have an undesired value
+                return false;
+            }
+
+            ValidateComparisonTerm(tokenToEvaluate);
+
+            // TODO throw this exception earlier?
+            if ((SpecifiedValue.Type == JTokenType.Date && tokenToEvaluate.Type != JTokenType.Date) ||
+                (tokenToEvaluate.Type == JTokenType.Date && SpecifiedValue.Type != JTokenType.Date))
+            {
+                throw new InvalidOperationException($"Cannot compare {SpecifiedValue.Type} with {tokenToEvaluate.Type} using an InequalityOperator");
```

Comment:
This should just evaluate to `false`.

Reply:
Why do you think this should evaluate to false? To avoid stopping the report of other results like Yanelis said? I'm more inclined to keep throwing an exception and catch it outside of the operator to show the information in the console but continue with the analysis, to give more information to the user, and also because returning false seems wrong in the context of the class. The Regex operator also throws an exception when trying to process an invalid regex. Should I create an issue to catch these exceptions?

Reply:
The Regex operator throws when interpreting the rule definition (the regex defined in the rule).  This is important because the rule can't evaluate anything if it's not defined properly.  That is different than throwing while evaluating a template, which is what's happening here.

Regardless of whether the two values are technically comparable or not, the intent is to simply discover places in a template that are inadequate or need to change.  If they aren't comparable, simply returning `false` will result in a rule failure, and the user will be notified of the failure and the line to look at.  This is cleaner and easier to understand than throwing an exception that needs to be caught later.

Reply:
But If the rule is written negating the output of this operator in some way? The user could potentially not find out about the issue. Or if the inequality operator restricts the search space like: where value > 10 check something else

Reply:
It could also be that the rule is expecting an int or something, but that property is never going to be an int

Reply:
@VeraBE, @yane3628, those are good examples where we might want something more than a simple pass/fail.  I think this does deserve more thought on how we handle these scenarios, and what information to surface to the user.

However, I don't think throwing an exception here is the right way to do it, and I don't think there's a pressing need to make issues like this highly visible just yet.  This should be designed in a more holistic way before implementing something like this.

Reply:
Okay, do you want me to add an issue for that? If we are going to return false for cases like that in the meantime, I feel the issue should have a pretty high priority, because we are risking many false negatives

Reply:
Yes, go ahead and create an issue for this.